### PR TITLE
Rework the Index and Quick Start guide

### PR DESCRIPTION
**Pull request by settermjd**
|    Q          |   A
|-------------- | ------
| Documentation | yes
| Bugfix        | no
| BC Break      | no
| New Feature   | no
| RFC           | no
| QA            | no

### Description

This intent of this PR isn't to turn the documentation into a marketing piece, rather ensure that it is a succinct, yet also sufficiently descriptive, introduction to what laminas-hydrator is. It's aimed at helping people who are seeing the package/component for the first time and need to quickly know what it is, and what they can do with it, and gives
them an incentive to delve deeper.

It:

1. Updates the package description to provide a more thorough introduction to the package
2. Reworks the Quick Start guide as follows:
  - Updates the introductory content to build upon the core description in index.md. This time, the user is a step deeper and seeking to learn more. So this description takes them that bit deeper, providing them quick and easy access to the next level down into the component.
  - Brings the usage section right up near the top, so that they can get a sample of how the package works and what they can expect. 
  - Moves the available implementations to after the usage section, simplifies the headers and descriptions of each one, and adds a usage example for any that don't have one. The reason for this is that, to me, using the FQN as the section header is redundant and too verbose. The descriptions were reworked to read a little more like a narrative, yet still maintain the focus of a core documentation guide. I tried hard to make the text flow, yet still, be active and to the point. There should be usage examples for each one, as you can never be sure which one the user may be most interested in, which one they need, and for completeness sake. I also sorted the hydrators into what seemed to be the most applicable order of complexity, from the most basic to the most advanced.

I need to check the usage examples for accuracy. After that, I'll remove the draft status.

### CODE REVIEW

File: docs/book/index.html
Comment (froschdesign) [inline docs/book/index.html]: This file is never used in the frontend and can be empty. Only MkDocs needs an index file to create the documentation.

File: docs/book/v3/quick-start.md
Comment (froschdesign) [inline docs/book/v3/quick-start.md]:
```suggestion
# Quick Start
```

File: docs/book/v3/quick-start.md
Comment (froschdesign) [inline docs/book/v3/quick-start.md]:
```suggestion
## Basic Usage
```

The quick start contains a "basic usage" therefore the headline should illustrate that, like in other components.

File: docs/book/v3/quick-start.md
Comment (froschdesign) [inline docs/book/v3/quick-start.md]: PHP's open and close tags are only allowed in code examples for view scripts.

File: docs/book/v3/quick-start.md
Comment (froschdesign) [inline docs/book/v3/quick-start.md]: Please remove PHP's open and close tags.

File: docs/book/v3/quick-start.md
Comment (froschdesign) [inline docs/book/v3/quick-start.md]: Not needed, will be create by MkDocs.

File: docs/book/v3/quick-start.md
Comment (froschdesign) [inline docs/book/v3/quick-start.md]:
```suggestion
$hydrator = new Laminas\Hydrator\ArraySerializableHydrator();
```

File: docs/book/v3/quick-start.md
Comment (froschdesign) [inline docs/book/v3/quick-start.md]:
```suggestion
$data = [
    'first_name'    => 'James',
    'last_name'     => 'Kahn',
    'email_address' => 'james.kahn@example.org',
    'phone_number'  => '+61 419 1234 5678',
];
```

See: https://docs.laminas.dev/laminas-coding-standard/v2/coding-style-guide/#2-7-arrays

File: docs/book/v3/quick-start.md
Comment (froschdesign) [inline docs/book/v3/quick-start.md]: This array is never used.

File: docs/book/v3/quick-start.md
Comment (froschdesign) [inline docs/book/v3/quick-start.md:37]: The title of the page is "Quick Start" and this section should extracted. Each hydrator should have its own page. (Please see also https://github.com/laminas/documentation/issues/13)

File: mkdocs.yml
Comment (froschdesign) [inline mkdocs.yml]: All descriptions should have the same content:

* Component documentation: https://docs.laminas.dev/laminas-hydrator/ (`mkdocs.yml`)
* Component overview: https://docs.laminas.dev/components/
* Packagist: https://packagist.org/packages/laminas/laminas-hydrator ([`composer.json`](https://github.com/laminas/laminas-hydrator/blob/3.2.x/composer.json#L3))
* GitHub repository: https://github.com/laminas/laminas-hydrator

File: docs/book/v3/quick-start.md
Comment (froschdesign) [inline docs/book/v3/quick-start.md]: Should be extracted to a separate page which describes the creation of custom hydrator otherwise the interfaces in the documentation are not useful.

**Review by froschdesign — CHANGES_REQUESTED** [review]
The page should be modified consistently and should show only a quick start. Everything else should be listed separately.

File: docs/book/index.html
Comment (settermjd) [reply docs/book/index.html]: Thanks for clarifying that. I'll get rid of it.

File: docs/book/v3/quick-start.md
Comment (settermjd) [reply docs/book/v3/quick-start.md]: I'll make that change locally.

File: docs/book/v3/quick-start.md
Comment (settermjd) [reply docs/book/v3/quick-start.md]: I'll make that change locally.

File: docs/book/v3/quick-start.md
Comment (settermjd) [reply docs/book/v3/quick-start.md]: Understood.

File: docs/book/v3/quick-start.md
Comment (settermjd) [reply docs/book/v3/quick-start.md]: Thanks for pointing that out.

File: docs/book/v3/quick-start.md
Comment (settermjd) [reply docs/book/v3/quick-start.md]: Good point. I still had to double-check the code examples, and saw that, along with some other issues.

File: docs/book/v3/quick-start.md
Comment (settermjd) [reply docs/book/v3/quick-start.md:37]: That's an excellent resource. I'm working on continuing the refactor of the docs to reflect it.

File: mkdocs.yml
Comment (settermjd) [reply mkdocs.yml]: You sure are thorough. I'm feeling fortunate to have you review my work.

File: docs/book/v3/quick-start.md
Comment (settermjd) [reply docs/book/v3/quick-start.md]: That's going to take a bit of time to complete, but I'll get it done.

File: mkdocs.yml
Comment (settermjd) [reply mkdocs.yml]: If you're happy with me making the change to the project's about info, I'll get it done, otherwise I'll wait until the PR's merged.

File: docs/book/v3/quick-start.md
Comment (settermjd) [reply docs/book/v3/quick-start.md]: This is going to take me quite some time to write. I might have to break this up in to several PRs

File: docs/book/v3/quick-start.md
Comment (settermjd) [reply docs/book/v3/quick-start.md:37]: I'm going to do that in a separate PR.

Comment (settermjd) [reply docs/book/v3/quick-start.md:37]: I've made a start, but it wil take a bit of time.

File: docs/book/v3/quick-start.md
Comment (settermjd) [reply docs/book/v3/quick-start.md]: I'm going to do that in a separate PR.

File: docs/book/v3/quick-start.md
Comment (settermjd) [reply docs/book/v3/quick-start.md]: I'll have to read more about MkDocs. Thanks for that.

File: mkdocs.yml
Comment (froschdesign) [reply mkdocs.yml]: Unfortunately the description is too long:

Bildschirmfoto 2020-09-23 um 07 49 45
Bildschirmfoto 2020-09-23 um 07 50 34

But I have another suggestion: we should add an introduction page where the description can be added. This would also for the "Quick Start" page because the quick start should explain the usage and not the component itself.

**Comment by settermjd** [comment]
@froschdesign, what do you think of the state of the PR as it currently stands?

File: composer.json
Comment (froschdesign) [inline composer.json]: Please revert this. We can find a shorter description in another pull request and add this sentence to the introduction also in another pull request.

File: docs/book/v3/quick-start.md
Comment (froschdesign) [inline docs/book/v3/quick-start.md]: Can be written in one line because the namespace is only used once:

```php
$hydrator = new Laminas\Hydrator\ArraySerializableHydrator();
```

File: docs/book/v3/quick-start.md
Comment (froschdesign) [inline docs/book/v3/quick-start.md]: Can be written in one line:

```php
$hydrator = new Laminas\Hydrator\ArraySerializableHydrator();
```

File: docs/book/v3/quick-start.md
Comment (froschdesign) [inline docs/book/v3/quick-start.md]: Please remove the open tag.

File: docs/book/v3/quick-start.md
Comment (froschdesign) [inline docs/book/v3/quick-start.md]: The same like before, the namespace is used only once therefore it can be reduced on one line.

File: docs/book/v3/quick-start.md
Comment (froschdesign) [inline docs/book/v3/quick-start.md]: Please remove.

File: docs/book/v3/quick-start.md
Comment (froschdesign) [inline docs/book/v3/quick-start.md]: The same like above: one line.

File: docs/book/v3/quick-start.md
Comment (froschdesign) [inline docs/book/v3/quick-start.md]: Please remove the open tag.

File: docs/book/v3/quick-start.md
Comment (froschdesign) [inline docs/book/v3/quick-start.md]: On one line.

File: docs/book/v3/quick-start.md
Comment (froschdesign) [inline docs/book/v3/quick-start.md]: Please remove.

File: docs/book/v3/quick-start.md
Comment (froschdesign) [inline docs/book/v3/quick-start.md]: Please remove this section because in the current version it does not help the reader.

File: mkdocs.yml
Comment (froschdesign) [inline mkdocs.yml]: Please revert this change.

**Review by froschdesign — CHANGES_REQUESTED** [review]
Only some small corrections.

**Comment by froschdesign** [comment]
@settermjd 
I added all open tasks to a project: https://github.com/laminas/laminas-hydrator/projects/1

**Pull request merged**
